% Template created by Karol Kozio≈Ç (www.karol-koziol.net) for ShareLaTeX

\documentclass[a4paper,9pt]{extarticle}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{graphicx}
\usepackage{xcolor}
\usepackage{tikz}

\usepackage{amsmath,amssymb,textcomp}
\everymath{\displaystyle}

\usepackage{times}
\renewcommand\familydefault{\sfdefault}
\usepackage{tgheros}
\usepackage[defaultmono,scale=0.85]{droidmono}

\usepackage{multicol}
\setlength{\columnseprule}{0pt}
\setlength{\columnsep}{20.0pt}


\usepackage{geometry}
\geometry{
a4paper,
total={210mm,297mm},
left=10mm,right=10mm,top=10mm,bottom=15mm}

\linespread{1.3}


% custom title
\makeatletter
\renewcommand*{\maketitle}{%
\noindent
\begin{minipage}{0.65\textwidth}
\begin{tikzpicture}
\node[rectangle,rounded corners=6pt,inner sep=10pt,fill=blue!50!black,text width= 0.95\textwidth] {\color{white}\Huge \@title};
\end{tikzpicture}
\end{minipage}
\hfill

\bigskip\bigskip
}%
\makeatother

% custom section
\usepackage[explicit]{titlesec}
\newcommand*\sectionlabel{}
\titleformat{\section}
  {\gdef\sectionlabel{}
   \normalfont\sffamily\Large\bfseries\scshape}
  {\gdef\sectionlabel{\thesection\ }}{0pt}
  {
\noindent
\begin{tikzpicture}
\node[rectangle,rounded corners=3pt,inner sep=4pt,fill=blue!50!black,text width= 0.95\columnwidth] {\color{white}\sectionlabel#1};
\end{tikzpicture}
  }
\titlespacing*{\section}{0pt}{15pt}{10pt}


% custom footer
\usepackage{fancyhdr}
\makeatletter
\pagestyle{fancy}
\fancyhead{}
\fancyfoot[C]{\footnotesize \textcopyright\ \@date\ \ \@author}
\renewcommand{\headrulewidth}{0pt}
\renewcommand{\footrulewidth}{0pt}
\makeatother


\title{Cortex for bacterial genomics: quickstart}
\date{2015}



\begin{document}

\maketitle

\begin{multicols*}{2}


\section{Requirements}

Download Cortex from github (https://github.com/iqbal-lab/cortex)  - there are download instructions there.

You must have installed VCFtools (and have the entire directory, not just the binary executable), R and Stampy. R must be in your path. Add the following directories to PERL5LIB and PATH
\begin{verbatim}
export PERL5LIB= /path/cortex/scripts/analyse_variants/
    bioinf-perl/lib;/path/cortex/scripts/calling:
    /path/VCFTools_dir/perl:$PERL5LIB
export PATH = /path/cortex/scripts/analyse_variants/
   needleman_wunsch-0.3.0
\end{verbatim}


\section{Preparation (once per species)}

There are a set of files Cortex needs to use, which you should not need to worry about, so we wrap it all up:
\begin{verbatim}
perl scripts/calling/prepare.pl --index INDEX  
     --ref_fa species.fasta 
     --dir_for_ref_objects /path/refdir/ 
     --vcftools_dir /path2/vcftools_1.0.9/
     --outdir /path3/results
     --stampy_bin /path4/stampy.py  --kmer 31
\end{verbatim}
This script will create a config script in /path3/results containing information about your parameter choices, so you do not need to enter them again in subsequent scripts.


\section{Compare samples against a reference; combine results; genotype}
{\bf Step1 }First we build per-sample graphs in parallel using commands such as this GNU parallels command. This will create a directory for each sample within directory /path3/results/.  A typical command would be (supposing there were 1700 samples):
\begin{verbatim}
parallel --gnu -j 20  
  perl scripts/calling/build_samples_parallel.pl 
  --num {}  --index INDEX  
  --outdir /path3/results --kmer 31 ::: {1..1700}
\end{verbatim}
{\bf Step2} Combine all the per-sample VCFs to get one combined set of sites (SNPs, indels, structural variants). 

\begin{verbatim}
perl scripts/analyse_variants/combine/combine_vcfs.pl 
 --prefix XYZ   --outdir  /path3/results --intersect_ref
   
\end{verbatim}
{\bf Step 3} Finally, independently genotype each sample at all of these sites. Here using GNU parallels to spread across 20 cores of a server. Memory use now drops as we only use the graph of poymorphisms.
\begin{verbatim}
cat /path3/results/combine/list_args_for_final_step | 
    parallel --colsep '\t'  perl scripts/calling/gt_1sample.pl
    --config /path3/results/combine/config.txt
    --invcf /path3/results/combine/XYZ.sites_vcf
    --sample {1} --outdir {2}  --sample_graph {3} 
\end{verbatim}
This will give you one VCF file per sample in \\
/path3/results/\{sample\_id\}/union\_calls/

\section{Segregating variants within our dataset (joint workflow)}
First build sample graphs as Step1 in the previous example. Then UNDETERMINED



\section{Pan-genome analysis}

To detect presence of a set of predefined genes (genes.fasta) among your samples
\begin{verbatim}
\end{verbatim}
To look at pan-genome graph of all samples and see which samples have which contigs, allowing you to stratify them by frequency or look for differentiating/segregating contigs.


\section{The End}

For further information:


\end{multicols*}

\end{document}
